File: algorithms/2_medium/B/00102_binary_tree_level_order_traversal/bfs.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> res{};

        if (root == nullptr) {
            return res;
        }

        queue<TreeNode*> queue{};

        queue.push(root);

        while (!queue.empty()) {
            const size_t level_size{queue.size()};
            vector<int> temp;
            temp.reserve(level_size);

            for (size_t i{0}; i < level_size; ++i) {
                TreeNode* node{queue.front()};
                queue.pop();
                temp.push_back(node->val);

                if (node->left != nullptr) {
                    queue.push(node->left);
                }
                if (node->right != nullptr) {
                    queue.push(node->right);
                }
            }

            res.emplace_back(std::move(temp));
        }

        return res;
    }
};

File: algorithms/1_easy/A/00637_average_of_levels_in_binary_tree/bfs.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def averageOfLevels(self, root: Optional[TreeNode]) -> List[float]:
        queue = deque()
        queue.append(root)
        res = []

        while queue:
            n = len(queue)
            total_sum = 0

            for i in range(n):
                node = queue.popleft()
                total_sum += node.val
                if node.left != None:
                    queue.append(node.left)
                if node.right != None:
                    queue.append(node.right)

            res.append(total_sum / n)

        return res
